File: Java/src/main/java/QuestionBank/Solution201_300/Solution236.java
package QuestionBank.Solution201_300;

import tool.TreeNode;

/**
 * @author windows
 */
public class Solution236 {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || p == root || q == root) {
            return root;
        }
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        if (left != null && right != null) {
            return root;
        }
        if (left == null){
            return right;
        }
        return left;
    }

    public static void main(String[] args) {
        Solution236 solution236 = new Solution236();
    }
}

File: C++/src/Solution201_300/Solution236.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution236.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
    if (root == nullptr || root == p || root == q){
        return root;
    }
    TreeNode* left = lowestCommonAncestor(root->left, p , q);
    TreeNode* right = lowestCommonAncestor(root->right, p, q);

    if (left && right){
        return root;
    }

    return left ? left : right;
}